python/pyspark/sql/connect/expressions.py

```diff
@@ -420,6 +420,12 @@ def __repr__(self) -> str:
             return f"{self._name}({', '.join([str(arg) for arg in self._args])})"
 
 
+class UnresolvedBinaryFunction(UnresolvedFunction):
```

Comment:
I think this issue is a bit complicated, since not all binary ops have this type of string repr, for example:

```
In [11]: df.name + 1
Out[11]: Column<'(name + 1)'>

In [12]: df.name + df.age
Out[12]: Column<'(name + age)'>

In [13]: df.name.startswith("n")
Out[13]: Column<'startswith(name, n)'>

In [14]: df.name.startswith(df.age)
Out[14]: Column<'startswith(name, age)'>
```

we may also need to take unary ops into account.

I guess we need to keep a list to control which function can be represented in this way.

Reply:
Is our goal to make the Column string representation consistent between client and server? We don't have a general rule about how to generate Column string representation, I think a lot of manual efforts are needed to check the server-side implementation and copy it to the client.

Shall we only take care of some common ones like attribute and literal?

Reply:
Since there was probably not a spec to follow for the string representation of these objects, we can try to do our best to make the doc tests pass and be somewhat compatible since this is not a real API:

  * binary ops with single character operator -> infix
  * other binary ops -> prefix
  * for other expression types we directly resolve check with the server

Reply:
ok, I think it is fine to start with only handling `binary ops with single character operator -> infix`.

I think we can simply modify the `__repr__` method of `UnresolvedFunction`, seems unnecessary to add a new class `UnresolvedBinaryFunction`

Reply:
I had this first but it feels weird and the code looks a bit ugly. in particular because the only place where we use this is in the Column class for the bin_op function. But given that we add even more logic I will change it back.

Reply:
I added the behavior for infix / prefix notation and fixed the alias behavior as well. For the remaining simple checks I did it looks good so far and should get us to a reasonable comparison.

I don't think the goal should be compatibility on this non-api but similar readability. If we find cases that are unreasonable to fix, we should apply the change that @HyukjinKwon proposed.